### ISSUE TITLE
Constructor compatibility for CrossHair symbolics

### ISSUE THREAD

**Issue by pschanely** [pschanely/CrossHair#94]
**Expected vs actual behavior**

CrossHair symbolic types cannot be constructed like their native counterparts are. (they take args like the z3 identifier for a symbolic, or a z3 expression). Mostly that's fine, but causes trouble when the code under analysis assumes it can take the type of a value, and construct a new value with the same type; e.g. `type(listarg)([1,2,3])`

This will create spurious errors, like in this [crosshair playground example](https://crosshair-web.org/?crosshair=0.1&python=3.8&gist=fe8098cb7c938bed3aa6d5008600837a&flags=verbose)

There are a couple of broad approaches for solving this problem, but I'm not ready to pull the trigger yet on those. One promising avenue is to patch `type()`, but this requires us to be very strict about what parts of the codebase should and shouldn't be patched - it's a bit of work.

**Comment by Rik-de-Kort** [comment]
Bit boneheaded maybe, but how about adding "vanilla" constructor capabilities to `SmtInt` and friends? Like evaluating `SmtInt(5)` gives you the same thing as `SmtInt(z3.IntVal(5))`.  
It would be a bit better than patching `type`, since (I think) we rely on the type machinery in various parts of the codebase.  
It'd also remove the stumbling block that both I and another dev game across.

**Comment by pschanely** [comment]
Not boneheaded at all - that's a viable approach. One gotcha is `SmtStr()` where a plain string argument means that it's creating a new symbolic with a z3 value having that name. Potentially, to work around this, we could migrate internal constructor usages to use special keyword arguments when creating using a z3 name. Happy to give you feedback on a more concrete proposal if you want to explore this.


Some background context about patching follows (which has been HEAVILY refactored with commits that I've pushed in the last few minutes!). This is just to give you an idea about how we deal with the fact that when running user code, we want symbolics to be invisible, and when we are inside CrossHair, we do not.

The [Patched()](https://github.com/pschanely/CrossHair/blob/a4f0822d81d95582f1be5c2dee0f6d8f1d44c839/crosshair/core.py#L119) context manager defines the outer scope that has patching applied. Patching can be temporarily disabled using the [NoTracing()](https://github.com/pschanely/CrossHair/blob/a4f0822d81d95582f1be5c2dee0f6d8f1d44c839/crosshair/tracers.py#L354) context manager (and re-enabled again later using [ResumedTracing()](https://github.com/pschanely/CrossHair/blob/a4f0822d81d95582f1be5c2dee0f6d8f1d44c839/crosshair/tracers.py#L362))

A LOT of things get patched; for example, `isinstance(SmtInt('x'), int)` returns True while patching is enabled.

**Comment by Rik-de-Kort** [comment]
After some thought, I think a classmethod is the most elegant way to go about it. `SymbolicStr.from_z3(s)` and `SymbolicStr.from_name('foo')`. Is quite alright. Then just put the straightforward constructor where it belongs IMO.

**Comment by pschanely** [comment]
Yup, that's a great idea. The classmethods still need to construct an instance somehow, but if you want to keep  `__init__` super clean, I think we can do `object.__new__(SymbolicStr)`.

**Comment by Rik-de-Kort** [comment]
Here's the pattern I'm trying out: initialize the object with a z3-based value, but to instantiate it with a Python value by overriding `__new__`. In the process converting `__init_var__` to `from_name`. Turns out it's a bit more complicated and there's a significant amount of instantiation logic I need to rewrite. No questions right now but figured I'd put a message anyway.

What do you mean with `object.__new__(SymbolicStr)`?

**Comment by pschanely** [comment]
Very exciting! Let me know how I can help.
Defining `__new__` sounds fine to me. You can also use `object.__new__` to get a completely un-initialized instance of a class; e.g.
```
>>> class Foo:
...   def __new__(cls, x):
...     assert False
...   def __init__(self, a, b):
...     assert False
>>> object.__new__(Foo)
<__main__.Foo object at 0x7fd64e21e240>
```

**Comment by pschanely** [comment]
I ended up needing to patch `type()` for other reasons anyway, so the original problem behind this issue is resolved as of 0.0.14.

Therefore, I'm going to mark the issue closed. That said, a lot of the code cleanup in #99 might still be pretty nice!
